services/jupyter/src/app/model/train_model.py

```diff
@@ -0,0 +1,324 @@
+        compute_metrics=compute_metrics,
+        train_dataset=train_encoded,
+        eval_dataset=val_encoded,
+        tokenizer=tokenizer,
+    )
+
+    trainer.train()
+
+    return trainer
+
+
+class WandbModelArtifact(BaseModel):
+
+    entity: str
+    project: str
+    artifact_name: str
+    tag: str
+
+    def __str__(self):
+        return f"{self.entity}/{self.project}/{self.artifact_name}:{self.tag}"
+
+
+NUM_LABELS = 4
+
+
+def upload_to_s3_bucket(
+    torchscript_dir: str = "model.pt",
+    bucket_name: str = "fsdl-model-test",
+    s3_filename: str = "model.pt",
+):
+    s3 = boto3.client(
+        "s3",
+        settings.AWS_REGION,
+        aws_access_key_id=settings.AWS_ACCESS_KEY_ID,
+        aws_secret_access_key=settings.AWS_SECRET_ACCESS_KEY,
+    )
+
+    # upload file from local directory to s3 bucket
+    with open(torchscript_dir, "rb") as f:
+        s3.upload_fileobj(f, bucket_name, s3_filename)
+
+
+def load_model_from_wandb(
+    artifact_params: WandbModelArtifact, return_dir: bool = False
+):
+
+    run = wandb.init()
+    artifact = run.use_artifact(str(artifact_params), type="model")
+    artifact_dir = artifact.download()
+
+    if return_dir:
+        return artifact_dir
+
+    model = AutoModelForSequenceClassification.from_pretrained(
+        artifact_dir,
+        num_labels=NUM_LABELS,
+        torchscript=True,
+    )
+
+    return model
+
+
+def test_routine(model: Optional[Any] = None):
+
+    model_ckpt = "distilbert-base-uncased"
+
+    _, _, test_dataset = load_data()
+
+    w = WandbModelArtifact(
+        entity="team_44",
+        project="model-registry",
+        artifact_name="distilbert-base-uncased-finetuned-news",
+        tag="v0",
+    )
+
+    if model is None:
+        load_model_from_wandb(w)
+
+    # test model
+    tokenizer = AutoTokenizer.from_pretrained(model_ckpt)
+
+    def tokenize(batch):
+        return tokenizer(batch["text"], padding=True, truncation=True)
+
+    test_encoded = test_dataset.map(tokenize, batched=True, batch_size=None)
+
+    results = test_model(model, test_encoded, tokenizer)
+
+
+def convert_model_to_torchscript(
+    model_dir: str, model_ckpt: str
+) -> torch.jit.ScriptModule:
+    """Convert a saved model into TorchScript using tracing."""
+
+    model = AutoModelForSequenceClassification.from_pretrained(
+        model_dir, num_labels=NUM_LABELS, torchscript=True
+    )
+
+    # Create example input with
+    # First row in 'ag_news' training dataset
+    dummy_input = {
+        "text": "Wall St. Bears Claw Back Into the Black (Reuters) Reuters - Short-sellers, Wall Street's dwindling\\band of ultra-cynics, are seeing green again.",
+        "label": 2,
+    }
+
+    # load tokenizer
+    tokenizer = AutoTokenizer.from_pretrained(model_ckpt)
+
+    dummy_tokenized_input = tokenizer(
+        dummy_input["text"], truncation=True, return_tensors="pt"
+    )
+
+    return torch.jit.trace(model, tuple(dummy_tokenized_input.values()))
```

Comment:
at some point we will need to migrate the CLI to the jupyter container to not rely on the docker host. I think it can stay since it's not being called.

Reply:
> we will need to migrate the CLI to the jupyter container to not rely on the docker host

Could you elaborate on why you think we should do that? :)

Reply:
the deploy commands should be accessed from a container. Using the CLI outside of one would force CI to clone the entire repo instead of pulling a docker image. 

I was thinking of using the `jupyter` container for all things model management